hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/compact/CompactFunction.java

```diff
@@ -74,14 +74,14 @@ public class CompactFunction extends ProcessFunction<CompactionPlanEvent, Compac
 
   public CompactFunction(Configuration conf) {
     this.conf = conf;
-    this.asyncCompaction = OptionsResolver.needsAsyncCompaction(conf);
+    this.asyncCompactionOperation = OptionsResolver.needsAsyncCompactionOperation(conf);
   }
```

Comment:
Why switching to a new option key, what's the purpose here ?

Reply:
We already support sync compaction for bounded source, it should be fine for async compaction for streaming source ?

Reply:
Normally we use compaction.async.enabled to turn on compaction. But we could not make it sync because it's already been true.
```java
    if (asyncCompaction) {
      // executes the compaction task asynchronously to not block the checkpoint barrier propagate.
      executor.execute(
          () -> doCompaction(instantTime, compactionOperation, collector, reloadWriteConfig()),
          (errMsg, t) -> collector.collect(new CompactionCommitEvent(instantTime, compactionOperation.getFileId(), taskID)),
          "Execute compaction for instant %s from task %d", instantTime, taskID);
    } else {
      // executes the compaction task synchronously for batch mode.
      LOG.info("Execute compaction for instant {} from task {}", instantTime, taskID);
      doCompaction(instantTime, compactionOperation, collector, writeClient.getConfig());
    }
```


> support sync compaction for bounded source

We will use sync compaction mode for unbounded source in some scenarios. And actually the bounded source sync compaction seems weird. It use compaction.async.enabled  true to turn on compaction, and then switch it to fasle for sync mode.

```java  
     // compaction
      if (OptionsResolver.needsAsyncCompaction(conf)) {   // here FlinkOptions.COMPACTION_ASYNC_ENABLED decides that we need compaction
        // use synchronous compaction for bounded source.
        if (context.isBounded()) {
          conf.setBoolean(FlinkOptions.COMPACTION_ASYNC_ENABLED, false); // we come here because it is true, and it's so weird to turn  it false, actually we just want the operation to be executed sync.
        }
        return Pipelines.compact(conf, pipeline);
      } else {
        return Pipelines.clean(conf, pipeline);
      }
```

Reply:
@danny0405 pls help see whether there are any more problems here?

Reply:
> And actually the bounded source sync compaction seems weird. It use compaction.async.enabled true to turn on compaction, and then switch it to fasle for sync mode.

I agree the logic here is a little weird, we can refactor it though,
the bounded source sync compaction is meaningful, especially when people do batch ingestion for mor table, they do not what a separate compaction job running again.

Reply:
Yes, agreed. 
Sync compaction is meaningful for both bounded and and unbounded source, because we sometimes will still use compaction in the ingestion job but not a separate compaction job for **unbounded** source: 
But we met cases that async compaction with watermark will cause compaction msgs missed. Thus we will keep sync compaction as an option.

Reply:
So, Why not jus add an option to indicate whether compaction is sync or not?
It can not only solve the problem that we cannot open sync compaction in ingestion job for unbounded source, but also help refactor the weird bounded source compaction logics.

Reply:
> But we met cases that async compaction with watermark will cause compaction msgs missed. Thus we will keep sync compaction as an option for unbounded source.

Can you elaborate more for the details here ?

Reply:
Watermark is also propagated by output.collector. 
In an ingestion job with async compaction and watermark, when upstream operators forward watermark,  and at the same time we use async thread to execute compaction async and collect the result compaction msgs,  the output.collector will become thread unsafe.

Reply:
> keep sync compaction as an option for unbounded source.

And in a whole for hudi feature, you might also agree that users should be provided sync compaction option for unbounded source, no matter the above mentioned scenarios?

Reply:
@danny0405 do we have any other questions here?

Reply:
> and at the same time we use async thread to execute compaction async and collect the result compaction msgs, the output.collector will become thread unsafe

The output collector is thread unsafe here, let's fix it to be thread safe then ~

Reply:
OK, but that will be another issue, i think. 
What about this pr ? We have discussed much about its necessity to keep sync compaction option for unbounded source.